File: mobile/android/app/src/main/kotlin/com/voquill/mobile/keyboard/KeyboardLayoutSpec.kt
package com.voquill.mobile.keyboard

import org.json.JSONArray
import org.json.JSONObject

data class KeyboardBottomRow(
    val mode: KeyboardKeySpec,
    val globe: KeyboardKeySpec,
    val space: KeyboardKeySpec,
    val delete: KeyboardKeySpec,
    val enter: KeyboardKeySpec,
)

data class KeyboardLayoutSpec(
    val languageCode: String,
    val alphaRows: List<List<KeyboardKeySpec>>,
    val numericRows: List<List<KeyboardKeySpec>>,
    val symbolRows: List<List<KeyboardKeySpec>>,
    val shift: KeyboardKeySpec,
    val bottomRow: KeyboardBottomRow,
) {
    companion object {
        fun fromJson(json: JSONObject): KeyboardLayoutSpec {
            val languageCode = json.optString("languageCode", "en")
            val alphaRows = parseRows(json.optJSONArray("alphaRows") ?: JSONArray())
            val numericRows = parseRows(json.optJSONArray("numericRows") ?: JSONArray())
            val symbolRows = parseRows(json.optJSONArray("symbolRows") ?: JSONArray())
            val shiftJson = json.optJSONObject("shift")
            val shift = if (shiftJson != null) parseKey(shiftJson) else KeyboardKeySpec(
                id = "shift", role = KeyboardKeyRole.SHIFT, label = "⇧"
            )
            val bottomRowJson = json.optJSONObject("bottomRow")
            val bottomRow = if (bottomRowJson != null) {
                KeyboardBottomRow(
                    mode = parseKey(bottomRowJson.optJSONObject("mode") ?: defaultModeKey()),
                    globe = parseKey(bottomRowJson.optJSONObject("globe") ?: defaultGlobeKey()),
                    space = parseKey(bottomRowJson.optJSONObject("space") ?: defaultSpaceKey()),
                    delete = parseKey(bottomRowJson.optJSONObject("delete") ?: defaultDeleteKey()),
                    enter = parseKey(bottomRowJson.optJSONObject("enter") ?: defaultEnterKey()),
                )
            } else {
                KeyboardBottomRow(
                    mode = parseKey(defaultModeKey()),
                    globe = parseKey(defaultGlobeKey()),
                    space = parseKey(defaultSpaceKey()),
                    delete = parseKey(defaultDeleteKey()),
                    enter = parseKey(defaultEnterKey()),
                )
            }
            return KeyboardLayoutSpec(
                languageCode = languageCode,
                alphaRows = alphaRows,
                numericRows = numericRows,
                symbolRows = symbolRows,
                shift = shift,
                bottomRow = bottomRow,
            )
        }

        fun parseRows(arr: JSONArray): List<List<KeyboardKeySpec>> {
            val rows = mutableListOf<List<KeyboardKeySpec>>()
            for (i in 0 until arr.length()) {
                val rowArr = arr.optJSONArray(i) ?: continue
                val row = mutableListOf<KeyboardKeySpec>()
                for (j in 0 until rowArr.length()) {
                    val keyObj = rowArr.optJSONObject(j) ?: continue
                    row.add(parseKey(keyObj))
                }
                rows.add(row)
            }
            return rows
        }

        fun parseKey(obj: JSONObject): KeyboardKeySpec {
            return KeyboardKeySpec(
                id = obj.optString("id", ""),
                role = KeyboardKeyRole.from(obj.optString("role", "")),
                label = obj.optString("label", ""),
                flex = obj.optInt("flex", 1),
                value = obj.optString("value").takeIf { it.isNotEmpty() },
            )
        }

        private fun defaultModeKey() = JSONObject().apply {
            put("id", "mode"); put("role", "mode"); put("label", "123"); put("flex", 1)
        }

        private fun defaultGlobeKey() = JSONObject().apply {
            put("id", "globe"); put("role", "globe"); put("label", "🌐"); put("flex", 1)
        }

        private fun defaultSpaceKey() = JSONObject().apply {
            put("id", "space"); put("role", "space"); put("label", "space"); put("flex", 4)
        }

        private fun defaultDeleteKey() = JSONObject().apply {
            put("id", "delete"); put("role", "delete"); put("label", "⌫"); put("flex", 1)
        }

        private fun defaultEnterKey() = JSONObject().apply {
            put("id", "enter"); put("role", "enter"); put("label", "return"); put("flex", 1)
        }
    }
}

File: mobile/ios/keyboard/Types/KeyboardLayoutSpec.swift
import Foundation

struct KeyboardBottomRow {
    let mode: KeyboardKeySpec
    let globe: KeyboardKeySpec
    let space: KeyboardKeySpec
    let delete: KeyboardKeySpec
    let enter: KeyboardKeySpec
}

struct KeyboardLayoutSpec {
    let languageCode: String
    let alphaRows: [[KeyboardKeySpec]]
    let numericRows: [[KeyboardKeySpec]]
    let symbolRows: [[KeyboardKeySpec]]
    let shift: KeyboardKeySpec
    let bottomRow: KeyboardBottomRow

    static func from(_ dict: [String: Any]) -> KeyboardLayoutSpec? {
        guard let languageCode = dict["languageCode"] as? String,
              let alphaRaw = dict["alphaRows"] as? [[Any]],
              let numericRaw = dict["numericRows"] as? [[Any]],
              let symbolRaw = dict["symbolRows"] as? [[Any]],
              let shiftDict = dict["shift"] as? [String: Any],
              let shift = KeyboardKeySpec.from(shiftDict),
              let bottomDict = dict["bottomRow"] as? [String: Any],
              let modeSpec = KeyboardKeySpec.from(bottomDict["mode"] as? [String: Any] ?? [:]),
              let globeSpec = KeyboardKeySpec.from(bottomDict["globe"] as? [String: Any] ?? [:]),
              let spaceSpec = KeyboardKeySpec.from(bottomDict["space"] as? [String: Any] ?? [:]),
              let deleteSpec = KeyboardKeySpec.from(bottomDict["delete"] as? [String: Any] ?? [:]),
              let enterSpec = KeyboardKeySpec.from(bottomDict["enter"] as? [String: Any] ?? [:])
        else { return nil }

        return KeyboardLayoutSpec(
            languageCode: languageCode,
            alphaRows: parseRows(alphaRaw),
            numericRows: parseRows(numericRaw),
            symbolRows: parseRows(symbolRaw),
            shift: shift,
            bottomRow: KeyboardBottomRow(
                mode: modeSpec,
                globe: globeSpec,
                space: spaceSpec,
                delete: deleteSpec,
                enter: enterSpec
            )
        )
    }

    private static func parseRows(_ arr: [[Any]]) -> [[KeyboardKeySpec]] {
        arr.compactMap { row in
            let specs = row.compactMap { item -> KeyboardKeySpec? in
                guard let dict = item as? [String: Any] else { return nil }
                return KeyboardKeySpec.from(dict)
            }
            return specs.isEmpty ? nil : specs
        }
    }
}
